### PR TITLE
Fix dialog button UX

### LINKED ISSUE #6230: Tab-index, pushing affordance, and hover styling for dialog buttons
**Describe the bug**
Insufficient styling for dialog buttons.

For example, the "File Changed" dialog (see #6229 for steps to trigger this dialog) does have tab-indexing, but it can only be seen (as of 0.35.4) in Firefox - chrome does not show an outline for the three options as you switch between them.

Example in Chrome (hitting tab does not show an outline on the buttons):
image

Same example in Firefox:
image

**Expected behavior**
All dialog buttons should contain a stylized tab-index indicator, a pushing affordance (some notion of button being clicked and triggering an action) and also hover styling for mouse-over indication of it being a clickable button.

- JupyterLab 0.35.4

**Related bugs**
#3473 is about having a tab-index and a sensible default, this bug is about styling the tab-index indicator

### PR DESCRIPTION
**Pull request by ellisonbg**
## References

Fixes #6230 

## Code changes

None.

## User-facing changes

* On dialog buttons only, use a 1px solid outline of our brand color 1, with an exterior offset of 4px as the focus indicator. This required making the widget have `overflow: visible` which isn't possible for all our buttons (in particular in toolsbars) as they have compact layout and this outline would be cut off.
* Slight opacity changes to indicate hover and active.
* Caps case for all buttons.
 
Screen Shot 2019-06-09 at 12 58 19 PM

Screen Shot 2019-06-09 at 12 58 28 PM

Screen Shot 2019-06-09 at 12 58 40 PM

Screen Shot 2019-06-09 at 12 58 49 PM

## Backwards-incompatible changes

None.

### CODE REVIEW

**Comment by blink1073** [comment]
The dialog test is looking for a capcase "cancel"

**Comment by blink1073** [comment]
Line 166

**Comment by ellisonbg** [comment]
Not sure why the tests are not passing....any insights?

**Comment by jasongrout** [comment]
I don't see the difference for the active case. For example, in this gif, I like the hover styling, but when I press and hold on the Select button, the border disappears, and I don't see any affordance that I'm actually pushing on the button. Did you mean to have some indication when a button was being pressed?
Dialog

**Comment by ellisonbg** [comment]
The active styling is subtle - I am removing the opacity change that happens on hover. A number of other UI toolkits are handling hover/active in this way (with incremental background changes).

**Comment by ellisonbg** [comment]
Rebased.

**Comment by jasongrout** [comment]
> The active styling is subtle - I am removing the opacity change that happens on hover. A number of other UI toolkits are handling hover/active in this way (with incremental background changes).

Can we keep the outline at least? Holding down the button makes it appear that the button loses focus since the outline disappears.

**Comment by jasongrout** [comment]
> with incremental background changes

Do they make it so that the incremental background change is exactly the same as the non-hover state? This design works well when there is a hover, but if I press the button using a keyboard, I won't see the transition from hover back to full opacity.

**Comment by ellisonbg** [comment]
It differs across different UI toolkits:

* Fabric gets darker on hover, darker still on active.
* Material Design does lighter on hover/focus (same lighter) and then the animation on press.
* Blueprint gets darker on hover, darker still on active.

But only blueprint shows a visual change on keypress. That isn't a css based styling (unless I am missing something) and these are not blueprint buttons so we don't get that.

I can move to the darker on hover, darker still on active if you want though.

**Comment by jasongrout** [comment]
Each of them distinguishes active from normal, which I do think is important.

**Comment by jasongrout** [comment]
Summary of a zoom conversation: for now, we're just making sure that the active style is different than the normal style.

Other things we realized:
* there are some differences in how buttons and dialogs work with keyboard handling among various ux systems. For example, does space on a button work on keydown or keyup? What about return? How about a button in a toolbar?

**Comment by ellisonbg** [comment]
After talking with @jasongrout I have made the hover and active styling get darker and darker still:

Normal:

Screen Shot 2019-06-10 at 6 17 15 PM

Hover:

Screen Shot 2019-06-10 at 6 17 35 PM

Active:

Screen Shot 2019-06-10 at 6 17 41 PM

**However** it now does this in dark mode too, but in dark mode the equivalent would be to get lighter. We can't do custom styling for the dark theme in the dialog because the dialog is mounted on the body, and the `data-theme-light` is on `.jp-LabShell`. I don't want to add new CSS theme variables as this problem is caused by problems in the theme variables and this would only pile more mess on.

**Comment by tgeorgeux** [comment]
@ellisonbg Originally I was going to ask if it makes sense to match the outline to the color of the button? IE: Red border on red buttons, blue on blue, grey on grey?

But when you look at the[ Material Design States](https://material.io/design/interaction/states.html#usage) for contained buttons (would apply to a modal) they have separate color stops for hover and focussed that change the color of the button, instead of adding a border. I think from a visual perspective this is more aesthetically pleasing, it's less crowded, and it looks more professional than adding the border. The one downside is that it might not have enough contrast change for accessibility. Thoughts?

Here's an example of buttons states in Material Design:
Screen Shot 2019-06-11 at 10 00 13 AM

**Comment by jasongrout** [comment]
@ellisonbg - now master has the theme data attribute on the body, so perhaps the dialog CSS can take advantage of that now, if you rebase?

**Comment by jasongrout** [comment]
@ellisonbg - I rebased  your changes on master so you could use the body-level data attributes.

In your current iteration, you hardcode what used to be the brand color as a blue. I tried a different approach using CSS filters that preserves the brand color:

```patch
diff --git a/packages/application/style/buttons.css b/packages/application/style/buttons.css
index f0d02156d..b7c980657 100644
--- a/packages/application/style/buttons.css
+++ b/packages/application/style/buttons.css
@@ -17,31 +17,39 @@ button {
 }
 
 button.jp-mod-styled.jp-mod-accept {
-  background: var(--md-blue-500);
+  background: var(--jp-brand-color1);
   border: 0;
   color: white;
 }
 
-button.jp-mod-styled.jp-mod-accept:hover {
-  background: var(--md-blue-600);
+[data-jp-theme-light='true'] button.jp-mod-styled.jp-mod-accept:hover,
+[data-jp-theme-light='true'] button.jp-mod-styled.jp-mod-reject:hover,
+[data-jp-theme-light='true'] button.jp-mod-styled.jp-mod-warn:hover {
+  filter: brightness(90%);
 }
 
-button.jp-mod-styled.jp-mod-accept:active {
-  background: var(--md-blue-700);
+[data-jp-theme-light='true'] button.jp-mod-styled.jp-mod-accept:active,
+[data-jp-theme-light='true'] button.jp-mod-styled.jp-mod-reject:active,
+[data-jp-theme-light='true'] button.jp-mod-styled.jp-mod-warn:active {
+  filter: brightness(80%);
 }
 
-button.jp-mod-styled.jp-mod-reject {
-  background: var(--md-grey-500);
-  border: 0;
-  color: white;
+[data-jp-theme-light='false'] button.jp-mod-styled.jp-mod-accept:hover,
+[data-jp-theme-light='false'] button.jp-mod-styled.jp-mod-reject:hover,
+[data-jp-theme-light='false'] button.jp-mod-styled.jp-mod-warn:hover {
+  filter: brightness(110%);
 }
 
-button.jp-mod-styled.jp-mod-reject:hover {
-  background: var(--md-grey-600);
+[data-jp-theme-light='false'] button.jp-mod-styled.jp-mod-accept:active,
+[data-jp-theme-light='false'] button.jp-mod-styled.jp-mod-reject:active,
+[data-jp-theme-light='false'] button.jp-mod-styled.jp-mod-warn:active {
+  filter: brightness(120%);
 }
 
-button.jp-mod-styled.jp-mod-reject:active {
-  background: var(--md-grey-700);
+button.jp-mod-styled.jp-mod-reject {
+  background: var(--md-grey-500);
+  border: 0;
+  color: white;
 }
 
 button.jp-mod-styled.jp-mod-warn {
@@ -50,14 +58,6 @@ button.jp-mod-styled.jp-mod-warn {
   color: white;
 }
 
-button.jp-mod-styled.jp-mod-warn:hover {
-  background: var(--md-red-600);
-}
-
-button.jp-mod-styled.jp-mod-warn:active {
-  background: var(--md-red-700);
-}
-
 .jp-Button-flat {
```

I don't quite like it because it makes the text darker in the light theme, where I really just want to change the background color. But there might also be some possibilities with using the contrast filter, for example.

**Comment by ellisonbg** [comment]
This also fixes our issues with multiple identical dialogs.

**Comment by ellisonbg** [comment]
Also closes #6589

**Comment by jasongrout** [comment]
Tests are actually passing, even if they don't show here: https://dev.azure.com/jupyterlab/jupyterlab/_build/results?buildId=2263

File: packages/apputils/src/dialog.ts [hunk 1]
```diff
@@ -43,18 +43,29 @@ export function showErrorMessage(
...
+  if (promise) {
+    return promise;
+  } else {
+    let dialogPromise = showDialog({
+      title: title,
+      body: body,
+      buttons: buttons
+    }).then(() => {
+      Private.errorMessagePromiseCache.delete(key);
+    });
```
Comment (vidartf) [inline packages/apputils/src/dialog.ts]: `catch` also?

Comment (jasongrout) [reply packages/apputils/src/dialog.ts]: Makes sense. Can you open an issue or PR?

**Pull request merged**